feat: DropDown 컴포넌트 제작

mizign opened the pull request:
## 어떤 변경인지 
- [x] Feat: 기능 변경, 기능 추가
- [ ] Docs: 문서 작업
- [ ] Refactor: 기능 변경 없이 코드 수정
- [ ] Style: 디자인
- [ ] Fix: 버그 수정
- [ ] Test: 테스트 코드 작성

## 설명
1. onChange prop은 필수로 할당해주세요
onChange prop으로 할당한 함수의 첫번째 파라미터로 **사용자가 선택한 옵션값이 string으로** 넘어갑니다
onChange 함수는 무조건 등록해서 사용해주셔야 DropDown 단에서 오류가 나지 않아요!

2. 선택된 option을 관리하는 state를 본 컴포넌트를 사용하는 페이지상에서 제작해주세요
![image](https://github.com/Codeit-Rolling-11-Letsgo/Rolling/assets/148179726/d188eb57-a861-4bec-92fd-46458ff5654d)
 option의 state를 바꾸는 setter함수를 onChange 함수에 넣어서 사용하시면 됩니다.

### 이슈 번호
>  예) .../Codeit-Rolling-11-Letsgo/Rolling/issues/**2** 
https://github.com/Codeit-Rolling-11-Letsgo/Rolling/issues/2

### 주요 변경 사항 

### 어떻게 동작하는지
option을 배열로, defaultValue와 errorMessage를 string으로, disabled값을 boolean 값으로 받아서 동작하는 컴포넌트입니다. (이래서 type을 명시하는 주석이 필요한거군요..!)
모두 button으로 처리했고, dropdown으로 내려올 옵션 버튼들은 placeholder쪽 버튼이 눌릴때마다 토글되도록 조건부로 처리했어요.

### To Reviewers
- prop으로 더 받아야할것이 없는지..? 컴포넌트가 이거 자체로 괜찮은건지 잘 모르겠습니다. 아직 '이 컴포넌트가 사용될때 고려할것'을 미리 생각해보는게 어렵네요.
- 일단 꼭 placeholder 쪽 버튼을 눌러야 드롭다운이 올라간다는 점이 좀 마음에 안드는데 개선 방안 추천 부탁드립니다..
 (placeholder버튼에 onBlur를 사용해봤었는데, 옵션 아래쪽 버튼을 눌렀을 때 onBlur 이벤트가 발생해서 옵션 버튼들에 onClick 이벤트가 발생하지 않더군요. 이런 것도 우선사항을 지정할 수 있을지 좀 궁금합니다)
- hover 이벤트 발생시 border가 2px → 1px로 줄어들어서 호버될 때 컴포넌트 크기 자체가 1px씩 줄어드는 것 같아요. 사용시 문제가 될까요?

Commit:
feat: DropDown 컴포넌트 제작

Commit:
feat: disabled, error 경우 대응

Commit:
fix: hover시 글자색 변경 적용

Commit:
fix: DropDown 폴더 생성

Commit:
docs: js doc 작성

mizign commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -3,6 +3,27 @@ import { useState } from 'react';
 import styles from '@/components/common//DropDown/DropDown.module.scss';
 import Icon from '@/components/common/icon/Icon';
 
+/**
+ * 옵션을 선택할 수 있는 드롭다운 컴포넌트입니다.
+ *
+ * @component
+ * @example
+ * // 기본 사용법:
+ * <DropDown options={['옵션 1', '옵션 2', '옵션 3']} defaultValue="옵션 선택" />
+ *
+ * // 에러 메시지가 있는 경우:
+ * <DropDown options={['옵션 1', '옵션 2', '옵션 3']} errorMessage="옵션을 선택해주세요" />
+ *
+ * @param {Object} props - 컴포넌트 속성들.
+ * @param {string[]} props.options - 드롭다운에 표시될 옵션들의 배열입니다.
+ * @param {string} [props.defaultValue='placeholder'] - 드롭다운의 기본값 또는 플레이스홀더입니다.
+ * @param {string} [props.errorMessage=''] - 에러 발생 시 표시할 메시지입니다.
+ * @param {string} [props.className] - 드롭다운 컨테이너에 적용할 추가 클래스 이름입니다.
+ * @param {boolean} [props.disabled=false] - 드롭다운을 비활성화할지 여부를 나타내는 부울 값입니다.
+ * @param {Object} [props...rest] - 나머지 프롭스를 컴포넌트의 루트 엘리먼트에 전파할 수 있습니다.
+ * @returns {JSX.Element} - 렌더링된 드롭다운 컴포넌트입니다.
+ */
+
```
컴포넌트 보여주고 Gpt 한테 작성해달라고 시켰더니 잘해주네요,,

dkile commented:
1. firefox에서 disabled 상태에서 내부 text와 아이콘 색상이 disabled 색상이 아니네요.
2. safari에서 스타일이 깨지는 이슈가 있네요. animation 때문인 것 같습니다.
<img width="1920" alt="스크린샷 2023-12-11 오후 3 00 36" src="https://github.com/Codeit-Rolling-11-Letsgo/Rolling/assets/29909393/140f3265-0cec-4486-a748-066631e32cb1">
3. hover 시에 border width 변화 때문에 레이아웃이 밀리는 현상이 있는데 이 부분은 https://stackoverflow.com/questions/19248985/how-to-prevent-shifting-when-changing-border-width
이걸 보시면서 해결하면 될 것 같습니다.
4. chrome에서도 width를 32rem으로 고정했지만 실제 크기는 480px로 나오네요.
5. 스타일도 border width 기본은 1px인데 focused와 actived 상태에만 2px이 되는 것 같습니다.

dkile commented on src/components/common/DropDown/DropDown.module.scss:
```diff
@@ -0,0 +1,77 @@
+	width: 32rem;
+	padding: 1.2rem 1.6rem;
+	display: flex;
+	flex-direction: row;
```
flex-direction은 기본 속성 값이 row이기 때문에 빼도 될 것 같습니다.

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,80 @@
+
+export default function DropDown({
+	options = ['옵션1', '옵션2', '옵션3'],
+	defaultValue = 'placeholder',
```
defaultValue 는 placeholder의 역할만 하고 실제로 옵션 중 하나의 값을 선택하게 동작하지 않는 것 같습니다. 제가 이해한 것이 맞나요?

mizign replied:
이거하고 위에 말씀해주신건 placeholder 버튼을 버튼 태그가 아니라 input태그의 button 타입으로 바꾸면 개선될 것 같아요. 아마 onChange도 사용하실 수 있을 듯한....? 사용자 입장에서 요론 개선점은 생각 못해봤는데 리뷰 감사합니다! 
++ 아닌것 같네요,, 이건 좀 더 고민해보겠습니다.

mizign replied:
handleSelect를 통해 옵션으로 받은 배열 값 중 하나를 선택할 수 있습니다!

ckswnskfk commented on src/components/common/DropDown/DropDown.module.scss:
```diff
@@ -0,0 +1,77 @@
+}
+
+.rotate {
+	transform: rotate(180deg);
```
간단하지만 멋진 애니메이션 효과네요 👍

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,80 @@
+import styles from '@/components/common//DropDown/DropDown.module.scss';
+import Icon from '@/components/common/icon/Icon';
+
+/**
+ * 옵션을 선택할 수 있는 드롭다운 컴포넌트입니다.
+ *
+ * @component
+ * @example
+ * // 기본 사용법:
+ * <DropDown options={['옵션 1', '옵션 2', '옵션 3']} defaultValue="옵션 선택" />
+ *
+ * // 에러 메시지가 있는 경우:
+ * <DropDown options={['옵션 1', '옵션 2', '옵션 3']} errorMessage="옵션을 선택해주세요" />
+ *
+ * @param {Object} props - 컴포넌트 속성들.
+ * @param {string[]} props.options - 드롭다운에 표시될 옵션들의 배열입니다.
+ * @param {string} [props.defaultValue='placeholder'] - 드롭다운의 기본값 또는 플레이스홀더입니다.
+ * @param {string} [props.errorMessage=''] - 에러 발생 시 표시할 메시지입니다.
+ * @param {string} [props.className] - 드롭다운 컨테이너에 적용할 추가 클래스 이름입니다.
+ * @param {boolean} [props.disabled=false] - 드롭다운을 비활성화할지 여부를 나타내는 부울 값입니다.
+ * @param {Object} [props...rest] - 나머지 프롭스를 컴포넌트의 루트 엘리먼트에 전파할 수 있습니다.
+ * @returns {JSX.Element} - 렌더링된 드롭다운 컴포넌트입니다.
+ */
```
저도 좀 참고해서 자세히 써봐야겠네요! 👍

ckswnskfk reviewed (commented):
수고하셨습니다! select 태그가 아니라 직접 만들어서 스타일링 하기 어려웠을 것 같은데 대단하십니다! 👍 

다만 기능적으로 어떻게 써야할지 잘 모르겠네요.
일반적으로 select 태그는 값을 사용자가 선택하고 그 값을 사용하기 위해 쓰는 것 같은데,
```jsx
<select value={val} onChange={handleChange}>
  <option value={'하나'}>하나</option>
  <option value={'둘'}>둘</option>
  <option value={'셋'}>셋</option>
</select>
```
그래서 이렇게 onChange 이벤트로 값을 핸들링 했던 것 같아요. 그런데 이 DropDown 컴포넌트에 onChange 이벤트를 줘도 값은 undefined 만 나와서 어떻게 사용할지 잘 모르겠네요 😢

또, 각 옵션을 클릭했을때와 컴포넌트 외부를 클릭했을 때 드롭다운이 닫히는 동작까지 구현되면 완벽할 것 같습니다!

Commit:
fix: 리뷰반영 (커밋 메세지 본문참조)

1. 파이어폭스, 사파리에서 일부 스타일이 적용되지 않는 오류 해결
2. onChange prop 추가
3. 일반/focus/active/hover 스타일 조정
4. focus시 border 두꺼워지며 컴포넌트 밀리는 현상 해결

Commit:
fix: 'disabled' prop default value 수정, 스타일 수정

Commit:
fix: 옵션 선택시 드롭다운 목록 닫히게

d0ngbb00 commented:
고생하셨어요 select 안 쓰고 구현하기 넘 어려웠을 것 같은데 jsdoc까지 깔끔하게 정리해주셔서 코드가 문제없이 읽히는 것 같아요🙌

ckswnskfk commented on src/components/common/DropDown/DropDown.module.scss:
```diff
@@ -0,0 +1,102 @@
+		color: $color_gray_300;
+		box-shadow: 0 0 0 0.1rem $color_gray_300;
+	}
+}
```
이 부분은 
```css
.selectContainer.disabled {
	box-shadow: 0 0 0 0.1rem $color_gray_300;
	background-color: $color_gray_100;
	color: $color_gray_300;

	&:hover {
		box-shadow: 0 0 0 0.1rem $color_gray_300;
	}
}
```
이렇게 되어야 호버 시 arrow icon이 깜빡거리지 않네요

ckswnskfk replied:
또 저처럼 disabled는 pseudo class로 넘기는 게 코드량도 줄고 깔끔해질 것 같습니다!

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,84 @@
+			>
+				<input
+					type='button'
+					onChange={onChange}
```
onChange 로 함수를 넘겨도 아무 동작하지 않는 것 같아요! console.log만 찍는 함수를 전달해도 실행이 안되는 걸 보니 onChange 이벤트가 일어나지 않는 것 같습니다.

mizign replied:
이게 placeholder 버튼의 value를 state로 바꾸고 있어서 값이 변경되면 재렌더링이 일어나서 onChange 이벤트는 감지가 안되는 듯합니다.. ㅎㅎ 이게 기능적으로 젤 중요한데 제일 어렵네요..

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,84 @@
+ */
+
+export default function DropDown({
+	options = ['옵션1', '옵션2', '옵션3'],
```
그런데 사용자가 보는 label과 개발자가 사용해야 할 값이 다른 경우는 어떻게 하나요?
ex)
```html
<select>
  <option value="dog">개</option>
  <option value="cat">고양이</option>
  <option value="hamster">햄스터</option>
</select>
```

ckswnskfk replied:
이건 요구사항을 보니 무시해도 될 것 같습니다!

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,84 @@
+	...props
+}) {
+	const [isArrowRotated, setArrowRotated] = useState(false);
+	const [placeholder, setPlaceholder] = useState(defaultValue);
```
결국 실제 값으로 쓰일 상태라고 보여지는데 placeholder 라는 이름이 적절한가 의문이 듭니다!

ckswnskfk reviewed (commented):
정말 어려운 공통 컴포넌트네요.. 그래도 잘 하실거라 믿습니다!

mizign commented:
> 정말 어려운 공통 컴포넌트네요.. 그래도 잘 하실거라 믿습니다!

꾸벅,, 감사드립니다

Commit:
fix: onChange 이벤트 발생하지 않던 오류 해결, state명 수정

Commit:
docs: dropDown js doc 수정

Commit:
refactor: clsx 사용, className 정돈

Commit:
refactor: disabled 상태 가상 클래스로 스타일링

dkile commented:
<img width="1710" alt="스크린샷 2023-12-12 오전 11 19 30" src="https://github.com/Codeit-Rolling-11-Letsgo/Rolling/assets/29909393/548a1f42-38d7-4c25-8a84-9d8c4df4cd3b">
option 값이 길면 overflow가 일어납니다. ellipsis나 overflow-hidden을 주면 좋을 것 같네요.

mizign commented:
> <img alt="스크린샷 2023-12-12 오전 11 19 30" width="1710" src="https://private-user-images.githubusercontent.com/29909393/289724772-548a1f42-38d7-4c25-8a84-9d8c4df4cd3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzNDg2MzMsIm5iZiI6MTcwMjM0ODMzMywicGF0aCI6Ii8yOTkwOTM5My8yODk3MjQ3NzItNTQ4YTFmNDItMzhkNy00YzI1LThhODQtOWQ4YzRkZjRjZDNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEyVDAyMzIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwYTRlM2NhNjYzMjRjNWJiMjY5MGJjNmNlYTY1MDA4ZDRhY2RmM2I2ZmQwMjhmMTUyOGYwNWNlMDVjNDJlNmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CVcWY1KNFKyPdM0nz7l-cPyXhnNdrmu-pIJ_HEhoZCk"> option 값이 길면 overflow가 일어납니다. ellipsis나 overflow-hidden을 주면 좋을 것 같네요.

ㅋㅋㅋㅋㅋㅋ아니 예시가 ㅋㅋ 수정하겠습니다.

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+
+	return (
+		<div className={className} {...props}>
+			<button
```
어제 멘토님 말씀하신 것 처럼 이 버튼에 `type='button'` 이게 들어가야겠네요!

mizign replied:
아앗 onChange 수정하고 여러번 리셋하다가 수정사항이 반영이 안됐네요.

Commit:
fix: 길이가 긴 옵션이 선택되었을 경우 옵션 텍스트가 컴포넌트 밖으로 나오는 현상 해결

dkile commented:
외부 클릭 시 닫는 기능은 나중에 구현할까요? 오늘 팀 회의 시간에도 얘기 하겠지만 이제 슬슬 페이지를 만들어야 할 것 같네요.

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+					{options.map((option) => (
+						<button
+							disabled={disabled}
+							key={option.slice(-1)}
```
option의 값에 의존해 키를 만드는 방법은 키가 중복될 위험이 있는것 같습니다!
ex) `options = ['축구', '야구', '배구']`
만 요구사항상 그럴 일은 없을 것 같긴 하네요ㅎㅎ

mizign replied:
uuid를 생성해서 key값을 지정하는 등의 방안도 물색해보았으나..? 이 방법도 리액트 동작원리상 좋지 않다고해요. 이런 경우 정말 키값을 어찌 해야하는지 고민을 많이했는데 일단 저희 프로젝트 상에서는 그럴일이 없다로 흘러가서 ,, 그냥 뒀습니다. 찬주님 QA 결과 반영 못한점 양해 부탁드려요 하핫

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+			{isArrowRotated && (
+				<ul className={styles.options}>
+					{options.map((option) => (
+						<button
```
`<li>` 태그가 있으면 더 좋을 것 같네요 😄

dkile commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+					{options.map((option) => (
+						<button
+							disabled={disabled}
+							key={option.slice(-1)}
```
key는 option.slice(-1)보다는 uuid를 사용해도 좋을 것 같네요.
option을 기반으로 하다보면 중복된 값이 존재할 수도 있을 것 같습니다.

mizign replied:
> uuid를 생성해서 key값을 지정하는 등의 방안도 물색해보았으나..? 이 방법도 리액트 동작원리상 좋지 않다고해요. 이런 경우 정말 키값을 어찌 해야하는지 고민을 많이했는데 일단 저희 프로젝트 상에서는 그럴일이 없다로 흘러가서 ,, 그냥 뒀습니다. 찬주님 QA 결과 반영 못한점 양해 부탁드려요 하핫

[uui](https://velog.io/@letthere/developic-react-list-key-error-uuid)

이 글을 읽고 그냥 뒀었는데, 어떻게 하면 좋을까요? 사실 사용 중에 dropdown 목록이 변하는 등의 일은 일어나지 않는다는 기획상의 요구에 따라서 이렇게 짰었거든요

dkile replied:
https://react.dev/learn/rendering-lists#where-to-get-your-key
공식문서에서 추천하는 방법은 uuid랑 crypto가 있네요

ckswnskfk commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+			<button
+				disabled={disabled || errorMessage}
+				className={clsx(
+					styles.selectContainer,
```
구조상 이 버튼에 `position: relative` 가 들어가서 options의 `position: absolute`가 상위 relative 요소를 찾지 못하고 body 에 붙어버리네요.
상위 div에 `position: relative`가 들어가면 좋을 것 같고 selectContainer 란 이름도 그곳에 더 어울리는 것 같습니다!

mizign replied:
핫 이부분도 수정을 했었는데 반영이 안됐네요,, 빗발치는 리뷰에 정신이 없었나 봅니다. 이부분은 left: inherit으로만 바꾸면 해결되더라구요.

ckswnskfk replied:
오 몰랐던 사실!

mizign commented:
> 외부 클릭 시 닫는 기능은 나중에 구현할까요? 오늘 팀 회의 시간에도 얘기 하겠지만 이제 슬슬 페이지를 만들어야 할 것 같네요.

이거는 
> placeholder버튼에 onBlur를 사용해봤었는데, 옵션 아래쪽 버튼을 눌렀을 때 onBlur 이벤트가 발생해서 옵션 버튼들에 onClick 이벤트가 발생하지 않더군요. 이런 것도 우선사항을 지정할 수 있을지 좀 궁금합니다

이 이슈때문에 옵션을 클릭하면 자동으로 닫히는 방안으로 변경했네요. 개선되면 좋겠지만 페이지 제작에 돌입해야한다는 점도 동의합니다~

dkile commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+	};
+
+	const handleSelect = ({ target }) => {
+		setSelectedOption(target.innerText);
```
target.innerText보다는 
const handleSelect = (option) => ({
    setSelectedOption(option);
})
onClick = {() => handleSelect(option)}
요런 식은 어떤가요?
혹은 value를 따로 지정하고 e.target.value를 사용하면 좋을 것 같네요.
만약 드롭다운 내에 아이콘 같은 추가 요소가 생긴다면 innerText는 문제가 발생할 수 있을 것 같습니다.

mizign replied:
placeholder 버튼쪽과의 일관성을 위해 button type input으로 변경하고 value 프로퍼티 사용하도록 하겠슴다

ckswnskfk reviewed (commented):
이제 외부에서 사용하는데 기능적으로는 큰 문제가 없을 것 같네요!
거의 다 온 것 같습니다ㅎㅎ 수고하셨어요!

dkile commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+			{errorMessage && (
+				<div className={styles.errorMessage}>{errorMessage}</div>
+			)}
+			{isArrowRotated && (
```
isArrowRotated란 이름보다는 isPopped나 isShown, isOpened 정도가 좋을 것 같습니다.
isArrowRotated는 기능과는 별로 상관이 없는 이름 같네요.

dkile commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+	const [isArrowRotated, setArrowRotated] = useState(false);
+	const [selectedOption, setSelectedOption] = useState(defaultValue);
+
+	useEffect(() => {
```
useEffect를 사용하기보다 handleSelect와 같은 핸들러 내에 onChange()를 두는 것을 추천드립니다.
useEffect는 사이드 이펙트를 처리할 때 사용한다고 생각하면 될 것 같습니다.

mizign replied:
그러면 onChange 함수가 사용자가 '드롭다운을 내려서 옵션을 선택'할때만 작동합니다. 시안을 보면 사용자가 dropdown을 내려 선택하지 않아도 기본으로 어떤 값이 선택되도록 구현해야해서 핸들러 함수가 아니라 useEffect 내의 콜백으로 배치했습니다..!

dkile replied:
혹시 defaultValue 말씀하시는건가요

dkile commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,92 @@
+		onChange(selectedOption);
+	}, [selectedOption, onChange]);
+
+	const handleArrowFocus = () => {
```
handleArrowFocus도 이름을 바꿔보면 좋을 것 같습니다.
arrow focus는 css 적인 요소라서 기능 상의 이름으로는 알아보기 어렵네요

Commit:
refactor: 리뷰사항 반영

mizign commented:
key 관련 제외하고 모두 리뷰 반영했습니다. 😇

Commit:
refactor: onChange 함수를 handleOptionSelect에 위치시킴

mizign commented:
용준님과 이야기 나눈 결과 선택된 option에 관한 state를 사용자 측에서 정의를 해두고 사용하게 하는 편이 더 낫다는 판단을 내렸습니다! 결론적으로 useEffect 구문은 삭제했습니다. 
사용하시는 분은 본문 설명 부분 참고해서 onChange 프롭에 setter함수를 할당해주세요

dkile commented on src/components/common/DropDown/DropDown.jsx:
```diff
@@ -0,0 +1,90 @@
+	...props
+}) {
+	const [isOpened, setIsOpened] = useState(false);
+	const [selectedOption, setSelectedOption] = useState(defaultValue);
```
그리고 state를 바깥에서 정의하면 내부의 state가 필요할지에 대해서도 한 번쯤 생각해볼 여지가 있습니다. 일단 지금은 기능 상으로는 문제가 없으니 상관 없지만 해당 state가 꼭 필요한가에 대해 한 번 생각해보면 좋을 것 같습니다.

mizign replied:
흠 근데 선택된 옵션의 텍스트로 바꿔서 재렌더링 하는 과정은 컴포넌트 내부에서 일어나야하니까 이 state도 필요하지 않나요??
